applications/FluidDynamicsApplication/custom_utilities/fluid_auxiliary_utilities.cpp

```diff
@@ -418,4 +421,75 @@ void FluidAuxiliaryUtilities::CalculateSplitConditionGeometryData<false>(
     rpModShapeFunc->ComputeNegativeExteriorFaceAreaNormals(rNormals, FaceId, GeometryData::IntegrationMethod::GI_GAUSS_2);
 }
 
+    void FluidAuxiliaryUtilities::EmbeddedSkinToVolumeMapping(
+        ModelPart& rVolumeModelPart,
+        ModelPart& rSkinModelPart,
+        const double SearchRadius
+        )
+    {
+        //put the nodes in the skin into a search database
+        typedef BinsDynamic<3, Node<3>, std::vector<Node<3>::Pointer>> DynamicBins;
+        DynamicBins locator(rSkinModelPart.Nodes().GetContainer().begin(), rSkinModelPart.Nodes().GetContainer().end());
+
+        //reset the EMBEDDED_VELOCITY
+        block_for_each(rVolumeModelPart.Nodes(), [](auto& rNode){
+            rNode.GetValue(EMBEDDED_VELOCITY).clear();
+        });
+
+        //detect all the nodes that belong to cut elements
+        ModelPart::NodesContainerType cut_elem_nodes;
+        for(const auto& rElem : rVolumeModelPart.Elements()){
+            const auto& geom = rElem.GetGeometry();
+            
+            unsigned int npos=0, nneg=0;
+            for(unsigned int i=0; i<geom.size(); ++i){
+                if(geom[i].FastGetSolutionStepValue(DISTANCE)>0)
+                    npos++;
+                else
+                    nneg++;
+            }
+
+            if(npos>0 && nneg>0){ //element is split
+                for(unsigned int i=0; i<geom.size(); ++i)
+                    cut_elem_nodes.push_back(geom(i));
+            }
+        }
+        cut_elem_nodes.Unique();
```

Comment:
i am not very clear you are right about this one specifically. The cost of what i have here is nlogn, and the cost of insering n times in a set should also be the same.

Reply:
point is, that later on a need to loop in parallel over this. Neither std::set nor std::unordered_set are suitable to that, so anyway i would need to copy back into a vector to be able to further process it in parallel. At that moment i prefer to push back to a vector and make it unique

Reply:
> i am not very clear you are right about this one specifically. The cost of what i have here is nlogn, and the cost of insering n times in a set should also be the same.

i correct myself. It is nlogn in a set, n in a unordered_set

Reply:
@matekelemen this is ready

Reply:
> later on a need to loop in parallel over this.

Good point. We should be able to partition any container that can be accessed by at least forward iterators. The STL can do it so it shouldn't be a problem for us either.

Reply:
to be honest i am not sure of how we can achieve this for an unorder_set without assumptions on the way the set is actually stored. (possibly stl does make those assumptions). In any case i agree that it would be interesting to have this capability...although of course not in this PR :-)